crates/llm-local/src/llama.rs

```diff
@@ -0,0 +1,168 @@
+use crate::{token_output_stream, utils::load_safetensors, CachedInferencingModel};
+use anyhow::{anyhow, Result};
+use candle::{utils, Device, Tensor};
+use candle_nn::VarBuilder;
+use candle_transformers::{
+    generation::{LogitsProcessor, Sampling},
+    models::llama::{self, Cache, Config, Llama, LlamaConfig},
+};
+use rand::{RngCore, SeedableRng};
+use spin_core::async_trait;
+use spin_world::v2::llm::{self as wasi_llm, InferencingUsage};
+use std::{fs, path::PathBuf, sync::Arc};
+use tokenizers::Tokenizer;
+
+const TOKENIZER_FILENAME: &str = "tokenizer.json";
+const CONFIG_FILENAME: &str = "config.json";
+const EOS_TOKEN: &str = "</s>";
+const MODEL_SAFETENSORS_INDEX: &str = "model.safetensors.index.json";
+
+pub fn auto_device() -> Result<Device> {
+    if utils::cuda_is_available() {
+        Ok(Device::new_cuda(0)?)
+    } else if utils::metal_is_available() {
+        Ok(Device::new_metal(0)?)
+    } else {
+        Ok(Device::Cpu)
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct LlamaModels {
+    model: Arc<Llama>,
+    config: Config,
+    cache: Cache,
+    tokenizer: Tokenizer,
+    device: Device,
+}
+
+impl LlamaModels {
+    pub fn new(model_dir: &PathBuf) -> Result<Self> {
+        let tokenizer_path = model_dir.join(TOKENIZER_FILENAME);
+        let config_path = model_dir.join(CONFIG_FILENAME);
+
+        let dtype = candle::DType::F16;
+        let device = auto_device()?;
+
+        let tokenizer = Tokenizer::from_file(tokenizer_path).map_err(|e| anyhow!(e.to_string()))?;
+        let config: LlamaConfig = serde_json::from_slice(&fs::read(config_path)?)?;
+
+        // TODO: flash attention is supposed to minimize memory read and writes - Do we want to turn it on
+        let config = config.into_config(false);
+        let cache = llama::Cache::new(true, dtype, &config, &device)?;
+
+        let safetensor_files = load_safetensors(&model_dir, MODEL_SAFETENSORS_INDEX)?;
+
+        // TODO: Check if there is a safe way to load the model from the file
+        let vb = unsafe { VarBuilder::from_mmaped_safetensors(&safetensor_files, dtype, &device)? };
+        let model = Llama::load(vb, &config)?;
+
+        Ok(Self {
+            model: Arc::new(model),
+            config,
+            cache,
+            tokenizer,
+            device,
+        })
+    }
+}
+
+#[async_trait]
+impl CachedInferencingModel for LlamaModels {
+    async fn infer(
+        &self,
+        prompt: String,
+        params: wasi_llm::InferencingParams,
+    ) -> anyhow::Result<wasi_llm::InferencingResult> {
+        let model = Arc::clone(&self.model);
+        let config = &self.config;
+        let tokenizer = self.tokenizer.clone();
+        let mut cache = self.cache.clone();
```

Comment:
Just conforming that we want a new `cache` everytime we run `infer`? I'm not sure how cache is supposed to be used here.

Reply:
As I understand it, we need a new cache for each `infer` call as the cache must contain only relevant data to each call.